File: Problemas/subset_sum.py
# Subset Sum Recursivo
def isSubsetSum(set, n, sum):

    if sum == 0: return True
    if n == 0 and sum != 0: return False
    if set[n-1] > sum:
        return isSubsetSum(set, n-1, sum)

    return isSubsetSum(set, n-1, sum) or isSubsetSum(set, n-1, sum-set[n-1])

# Subset Sum DP
def subsetSum(st, n, sm):
    subset = [[True for _ in range(sm+1)] for _ in range(n+1)]
    for i in range(n+1): # if sum == 0
        subset[i][0] = True
    
    for i in range(1, sm+1): # if n == 0 and sum != 0
        subset[0][i] = False
    
    for i in range(1, n+1):
        for j in range(1, sm+1):
            if j < st[i-1]: # if set > sum
                subset[i][j] = subset[i-1][j] 
            else:
                subset[i][j] = subset[i-1][j] or subset[i-1][j - st[i-1]]
    return subset[n][sm]

File: Problemas/subset_sum.cpp
#include <vector>
using namespace std;
typedef vector<int> vi;

bool isSubsetSum(int coins[], int soma, int M)
{
    bool *possible = new bool[soma + 1]();
    possible[0] = true;
    for (int c = 0; c < M; c++)
    {
        for (int i = soma - coins[c]; i >= 0; i--)
        {
            if (possible[i])
            {
                possible[i + coins[c]] = true;
            }
        }
    }
    return possible[soma];
}


vi moedas;

bool isSubsetSum(int sum){
    vector<bool> possible(sum+1, false);
    possible[0] = true;
    for(auto c : moedas)
        for(int i = sum-c; i >= 0; i--)
            if(possible[i])
                possible[i+c] = true;
    return possible[sum]; 
}